### ISSUE TITLE
File not cached problem if the permalink is non-english characters.

### ISSUE THREAD

**Issue by cheonmux** [SatelliteWP/rocket-nginx#179]
**Describe the bug**
File not cached problem if the permalink is non-english characters.

**Versions**
What version of Nginx are you using ?   1.12.6
What version of Rocket-Nginx are you using ? 3.0 
What version of WP Rocket are you using ?  3.11.0.5
Are you using Nginx as a reverse proxy (with Apache for instance) ? No

Did you activate the debug in Rocket-Nginx ? Please do and include any headers.

Not cached page
```
HTTP/1.1 200 OK =>
Server => nginx
Date => Tue, 03 May 2022 16:03:40 GMT
Content-Type => text/html; charset=UTF-8
Connection => close
Vary => Accept-Encoding
Last-Modified => Tue, 03 May 2022 08:25:06 GMT
X-Rocket-Nginx-Serving-Static => MISS
X-Rocket-Nginx-Reason => File not cached
X-Rocket-Nginx-File => /var/www/ja.seoartgallery.com/htdocs/wp-content/cache/wp-rocket/ja.seoartgallery.com//product/%25e3%2581%25b2%25e3%2581%25be%25e3%2582%258f%25e3%2582%258a-%25e7%25b5%25b5-%25e9%25a2%25a8%25e6%25b0%25b4-%25e7%25b5%25b5%25e7%2594%25bb-%25e5%25a3%2581%25e6%258e%259b%25e3%2581%2591-1///index-https.html
```

Cached page
```
HTTP/1.1 200 OK =>
Server => nginx
Date => Tue, 03 May 2022 16:09:10 GMT
Content-Type => text/html; charset=utf-8
Content-Length => 506530
Last-Modified => Tue, 03 May 2022 08:26:14 GMT
Connection => close
Vary => Accept-Encoding, Cookie
ETag => "6270e726-7baa2"
Cache-Control => no-cache, no-store, must-revalidate
X-Rocket-Nginx-Serving-Static => HIT
X-Rocket-Nginx-Reason => /wp-content/cache/wp-rocket/ja.seoartgallery.com//terms-of-service///index-https.html
X-Rocket-Nginx-File => /var/www/ja.seoartgallery.com/htdocs/wp-content/cache/wp-rocket/ja.seoartgallery.com//terms-of-service///index-https.html
Accept-Ranges => bytes
```

Make sure you include ALL Nginx configuration files you are using. Remove any sensitive information before submitting.

```
map $http_accept $suffix {
~image/webp    .webp;
default        '';
	}

server {
    listen       80;
    server_name   ja.seoartgallery.com www.ja.seoartgallery.com;
 
    return       301 https://$server_name$request_uri;
}


server {
    listen       443 ssl http2;
	listen [::]:443 ssl http2;

    server_name  ja.seoartgallery.com www.ja.seoartgallery.com;
    root /var/www/ja.seoartgallery.com/htdocs;
 
    index  index.php index.html index.htm;

	ssl_certificate /etc/letsencrypt/live/ja.seoartgallery.com/fullchain.pem;
	ssl_certificate_key /etc/letsencrypt/live/ja.seoartgallery.com/privkey.pem;
    ssl_dhparam /etc/ssl/certs/dhparam.pem;
	ssl_trusted_certificate /etc/letsencrypt/live/ja.seoartgallery.com/chain.pem;	
	ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3;

    # Set caches, protocols, and accepted ciphers. This config will merit an A+ SSL Labs score.
    #ssl_session_cache shared:SSL:10m;
   ssl_session_timeout 5m;
    ssl_prefer_server_ciphers on;
	ssl_ciphers TLS13-CHACHA20-POLY1305-SHA256:TLS13-AES-256-GCM-SHA384:TLS13-AES-128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384;

   # ssl_ciphers 'ECDH+AESGCM:ECDH+AES256:ECDH+AES128:DH+3DES:!ADH:!AECDH:!MD5';
	ssl_ecdh_curve X25519:prime256v1:secp521r1:secp384r1;
	ssl_session_tickets off;
	
	ssl_stapling on; 
	ssl_stapling_verify on; 
	ssl_buffer_size 4k;


    access_log off;
    error_log  /var/log/nginx/ja.seoartgallery.com.error.log;


	# START Nginx Rewrites for Rank Math Sitemaps
	rewrite ^/sitemap_index.xml$ /index.php?sitemap=1 last;
	rewrite ^/([^/]+?)-sitemap([0-9]+)?.xml$ /index.php?sitemap=$1&sitemap_n=$2 last;
	# END Nginx Rewrites for Rank Math Sitemaps


  location ~ /wp-content/(?<path>.+)\.(?<ext>jpe?g|png|gif)$  {
		add_header Vary Accept;
		expires 365d;
	   try_files /wp-content/uploads-webpc/$path.$ext$suffix /wp-content/$path.$ext$suffix $uri =404;
	}
	 location ^~ /.well-known/apple-developer-merchantid-domain-association {
			auth_basic off;
			allow all;
			try_files $uri =404;
			break;
		}
 
     client_max_body_size 100M;



    location / {
        try_files $uri $uri/ /index.php?$args; 
		add_header Access-Control-Allow-Origin *;
    }		

    # Add PHP handler 이거 LAEL 설정임. 
    location ~ [^/]\.php(/|$) {
		fastcgi_split_path_info ^(.+?\.php)(/.*)$;
		if (!-f $document_root$fastcgi_script_name) {
			return 404;
		}
		try_files $uri =404;

		fastcgi_read_timeout 30;
		fastcgi_pass unix:/run/php/php7.4-fpm.sock;
		fastcgi_index index.php;
		fastcgi_buffers 64 16k; # default 8 4k
		include fastcgi_params;

    }	
		
	include rocket-nginx/conf.d/default.conf;
	include optimization/*.conf;
}

```
```

user  www-data;
worker_processes  1;

error_log  /var/log/nginx/error.log crit;
pid        /var/run/nginx.pid;

worker_rlimit_nofile 10000; 
events {
    worker_connections  1024;
	multi_accept off;
}


http {
	charset           utf-8;
		
    include       /etc/nginx/mime.types;
    default_type  application/octet-stream;



	#DDOS 보안	
	# limit the number of connections per IP
	limit_conn_zone $binary_remote_addr zone=conn_limit_per_ip:10m;

	# limit the number of requests for this session
	limit_req_zone $binary_remote_addr zone=req_limit_per_ip:10m rate=10r/s;

    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log off;

    sendfile        on;
    tcp_nopush     on;
	tcp_nodelay on;
	server_tokens off;
	client_max_body_size 64m;
	add_header X-UA-Compatible "IE=Edge,chrome=1";

	#타임아웃 설정 
	#https://www.digitalocean.com/community/tutorials/how-to-optimize-nginx-configuration
	client_body_timeout 14;
	client_header_timeout 14;
	keepalive_timeout 25;
	send_timeout 13;


    #gzip  on;
	#오픈파일캐쉬
	# 이걸로 다시 테스트
	open_file_cache max=1500 inactive=30s;
	open_file_cache_valid    60s;
	open_file_cache_min_uses 2;
	open_file_cache_errors   on;
	
	
	brotli on;
	brotli_comp_level 6;
	brotli_static on;
	brotli_types application/atom+xml application/javascript application/json application/rss+xml
	   application/vnd.ms-fontobject application/x-font-opentype application/x-font-truetype
	   application/x-font-ttf application/x-javascript application/xhtml+xml application/xml
	   font/eot font/opentype font/otf font/truetype image/svg+xml image/vnd.microsoft.icon
	   image/x-icon image/x-win-bitmap text/css text/javascript text/plain text/xml;
		include /etc/nginx/conf.d/*.conf;
}


```

**Expected behavior**

hello.

I run the website in Japanese, and most of the permalinks are written in Japanese.

Recently, after enabling the debug of rocket-nginx, I found the file not being cached.

And I've found that this problem only occurs when the permalink is written as a non-english characters.

If there is a permalink called 絵画, it is encoded as %e7%b5%b5%e7%94%bb and saved in the /wp-content/cache/wp-rocket/ja.seoartgallery.com/%e7%b5% b5%e7%94%bb/ folder.

I think this problem is caused by the string difference between 絵画 and %e7%b5%b5%e7%94%bb.

I reported this to the wp-rocket team, but the rocket team got a response saying to report rocket-nginx.

Can you please let me know if there is a solution?

**Comment by maximejobin** [comment]
Did you get an answer from the WP Rocket team?

The folder is created by WP Rocket. The URL is handled by Nginx.

I'll wait for your reply.

**Comment by cheonmux** [comment]
Hi, @maximejobin 

Thanks for answering. I will wait for this issue to be resolved. :)

**Comment by medtr-git** [comment]
Hi,
It appears that the problem hasn't been resolved yet.
I can confirm the observation made by @cheonmux, wherein URLs containing non-English characters consistently yield a 'MISS' response with the reason being 'File not cached'.

Here are the response headers for your reference:
```
Cache-Control:no-cache, must-revalidate

Date:Wed, 20 Mar 2024 23:57:44 GMT

Expires:Wed, 20 Mar 2024 23:57:44 GMT

Last-Modified:Wed, 20 Mar 2024 23:57:31 GMT

Server:nginx

Strict-Transport-Security:max-age=31536000;

X-Rocket-Nginx-File:/home/admin/web/****/public_html/wp-content/cache/wp-rocket/****/product/%25d8%25ac%25d8%25b2%25d8%25a7%25d8%25b2%25d8%25a9-%25d8%25a7%25d9%2584%25d8%25b9%25d8%25b4%25d8%25a8///index-https.html

X-Rocket-Nginx-Reason:File not cached

X-Rocket-Nginx-Serving-Static:MISS
```

Here's the correct path where the cached file is stored:
`/home/admin/web/****/public_html/wp-content/cache/wp-rocket/****/product/%d8%ac%d8%b2%d8%a7%d8%b2%d8%a9-%d8%a7%d9%84%d8%b9%d8%b4%d8%a8/`

The issue is the result the use of '%' encoding which is '%25' by Rocket Nginx when attempting to access the file, while the directory where the file is saved by WP Rocket utilizes the standard version '%'. This inconsistency leads to the problem encountered.

@maximejobin, I hope this information helps in resolving the issue. Thank you

**Comment by maximejobin** [comment]
I understand the issue but I'm unable to reproduce it.

What language is used on your website @medtr-git ?

**Comment by medtr-git** [comment]
I'm using Arabic language.

To reproduce the issue, please create a product with the title 'جزازة العشب' (this is the exact product title I used in the previous reply).

Looking forward to your response.

**Comment by maximejobin** [comment]
After lots of trials and errors, I finally can say this won't be possible.

I had found a way to stop the double encoding. The latest problem is that the browser sends the filename as "%D8" (uppercase) but the file is saved as "%d8" (lowercase) by WP Rocket. There is no way to change the case without installing a third party module. I also looked at the RFC and it seems like it should be case insensitive so both "%d8" and "%D8" can be expected from the browser.

Based on that, it is a limitation and it won't be fixed... unless a future version of Nginx adds a way to solve this issue without needing a third party module.

I'm sad about this answer, considering the amount of time spent on this, but it is what it is...

**Comment by maximejobin** [comment]
I just released the beta version for Rocket-Nginx 3.1.0.

I've added a "preprocess" include where you could add code linked to a third party module (like Perl or Lua) to solve your issue. That way, it can be solved and Rocket-Nginx does not need a dependency!

See the Limitations section about "encoded URLs" for more information:
https://github.com/SatelliteWP/rocket-nginx/blob/version-3.1/README.md#limitations

Here is the link to the beta version:
https://github.com/SatelliteWP/rocket-nginx/releases/tag/v3.1.0-beta

Enjoy!
